### PR TITLE
Proof generation and verification.

### PR DESCRIPTION
**Pull request by chriseth**
This is the completed implementation of the full proof generation and verification plus tests that verification is correct for the generated proofs and additional tests that random one-nibble changes in the proofs do not verify.

Please merge #9  first.

### CODE REVIEW

**Comment by terminaldweller** [comment]
👍 
that's a lotta code. gimme a couple hours to look through.

File: contracts/scryptFramework.sol [hunk 1]
```diff
@@ -24,6 +25,47 @@ contract ScryptFramework {
         bytes32[] proof;
     }
 
+    // Only used for testing.
+    function hashState(State memory state) pure internal returns (bytes32) {
+        return keccak256(state.memoryHash, state.vars, state.inputHash);
+    }
+    function encodeState(State memory state) pure internal returns (bytes r) {
```
Comment (chriseth) [inline contracts/scryptFramework.sol:32]: This are the main encoding and decoding functions.

File: test/index.js [hunk 2]
```diff
@@ -108,22 +155,6 @@ var expectations = {
     ]
 }
 
-async function deployIfNeeded(account) {
```
Comment (chriseth) [inline test/index.js:111]: Not needed anymore.

File: test/index.js [hunk 3]
```diff
@@ -161,61 +222,92 @@ async function testStepValues(account) {
...
-        console.log("success".green)
-    } else {
-      //process.exit(1)
+    if (result.output != "0xda26bdbab79be8f5162c4ca87cc52d6f926fb21461b9fb1c88bf19180cb5c246") {
+        console.log("Invalid result after step 2049: ".red + result.output)
+        anyError = true
     }
 }
 
-async function verifyInnerStepTest(account) {
```
Comment (chriseth) [inline test/index.js:175]: This is superseeded by the generic step verifier.

**Comment by terminaldweller** [comment]
apparently i cant directly commit changes to a pull request. i made a new pull request with the changes...
so maybe close this and just merge the new one?

**Pull request closed without merging**